File: src/gui/itemnode/StringSplitGenerator.java
package gui.itemnode;

import gui.itemnode.ItemNode.ValueNode;
import javafx.scene.control.TextField;

public class StringSplitGenerator extends ValueNode<StringSplitParser> {
	private final TextField root = new TextField();

	public StringSplitGenerator() {
		root.setPromptText("expression");
		root.textProperty().addListener((o, ov, nv) -> generateValue(nv));
		generateValue("%Out%");
	}

	@Override
	public TextField getNode() {
		return root;
	}

	private void generateValue(String s) {
		try {
			root.setText(s);
			changeValue(new StringSplitParser(s));
		} catch (IllegalArgumentException e) {}    // ignore invalid values
	}

}

File: src/player/main/sp/it/pl/gui/itemnode/StringSplitGenerator.kt
package sp.it.pl.gui.itemnode

import javafx.scene.control.TextField
import sp.it.util.reactive.attach
import sp.it.util.text.StringSplitParser

class StringSplitGenerator: ValueNode<StringSplitParser>(StringSplitParser.singular()) {
    private val node = TextField()

    init {
        node.promptText = "expression"
        node.text = value.expression
        node.textProperty() attach { generateValue(it) }
    }

    override fun getNode() = node

    private fun generateValue(s: String) {
        StringSplitParser.fromString(s).ifOk(this::changeValue)
    }

}